plugins/net_plugin/net_plugin.cpp

```diff
@@ -2059,33 +2053,33 @@ namespace eosio {
    // thread safe
    bool dispatch_manager::add_peer_block( const block_id_type& blkid, uint32_t connection_id) {
       std::lock_guard<std::mutex> g( blk_state_mtx );
-      auto bptr = blk_state.get<by_id>().find( std::make_tuple( connection_id, std::ref( blkid )));
+      auto bptr = blk_state.get<by_connection_id>().find( std::make_tuple(std::ref(blkid), connection_id) );
       bool added = (bptr == blk_state.end());
       if( added ) {
-         blk_state.insert( {blkid, block_header::num_from_id( blkid ), connection_id, true} );
-      } else if( !bptr->have_block ) {
-         blk_state.modify( bptr, []( auto& pb ) {
-            pb.have_block = true;
-         });
+         blk_state.insert( {blkid, block_header::num_from_id( blkid ), connection_id} );
       }
       return added;
    }
 
    bool dispatch_manager::peer_has_block( const block_id_type& blkid, uint32_t connection_id ) const {
       std::lock_guard<std::mutex> g(blk_state_mtx);
-      const auto blk_itr = blk_state.get<by_id>().find( std::make_tuple( connection_id, std::ref( blkid )));
+      const auto blk_itr = blk_state.get<by_connection_id>().find( std::make_tuple(std::ref(blkid), connection_id) );
       return blk_itr != blk_state.end();
    }
 
    bool dispatch_manager::have_block( const block_id_type& blkid ) const {
       std::lock_guard<std::mutex> g(blk_state_mtx);
-      // by_peer_block_id sorts have_block by greater so have_block == true will be the first one found
-      const auto& index = blk_state.get<by_peer_block_id>();
+      const auto& index = blk_state.get<by_connection_id>();
       auto blk_itr = index.find( blkid );
```

Comment:
See https://www.boost.org/doc/libs/1_81_0/libs/multi_index/doc/tutorial/key_extraction.html search for "find" and read about composite keys.

Reply:
Maybe I don't understand. How could `index.count( blkid ) > 0` be faster than `index.find( blkid )`?  `find()` just has to find the first one, `count()` has to count how many are in the range. We have potentially a block id for each connection so that is very likely to be `> 1`.

Reply:
You are right. I really have to read more about boost multi-index. Sorry for all the noise!

Reply:
No noise at all. It was good to double check.